### PR TITLE
Expose the engine id when registering and show the engine details for list engine command

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Expose the engine id when registering the service. 

So that we can get more details about the engine when using `list engine` command.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
```
AdminResourceSuite:
- refresh Hadoop configuration of the kyuubi server
- delete engine - user share level *** FAILED ***
  List("serviceUri=localhost:40419;version=1.7.0-SNAPSHOT;kyuubi.engine.id=local-1666594567035;refId=823f7053-4ec6-4eaf-8521-3dc209393673;sequence=0000000000") had size 1 instead of expected size 0 (AdminResourceSuite.scala:103)
```

**Comment by yaooqinn** [comment]
why a engine id means more information？

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala [hunk 1]
```diff
@@ -342,9 +351,12 @@ class ZookeeperDiscoveryClient(conf: KyuubiConf) extends DiscoveryClient {
 
     val session = conf.get(HA_ENGINE_REF_ID)
       .map(refId => s"refId=$refId;").getOrElse("")
+    val extraInfo =
+      extraServiceInfo.map(kv => kv._1 + "=" + kv._2).mkString(";", ";", "").stripSuffix(";")
     val pathPrefix = ZKPaths.makePath(
       namespace,
-      s"serviceUri=$instance;version=${version.getOrElse(KYUUBI_VERSION)};${session}sequence=")
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala:347]: what else will be added to to path?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala [hunk 2]
```diff
@@ -92,6 +92,10 @@ class SparkTBinaryFrontendService(
       None
     }
   }
+
+  override def extraServiceInfo: Map[String, String] = {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala:96]: why is spark special?

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala:96]: image

Now I can only see the `KYUUBI_ENGINE_ID` is used in Spark.

We can extend other engines to support expose kyuubi.engine.id.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala (hunk 1 above)
Comment (turboFei) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala:347]: now only the kyuubi.engine.id is added.

Maybe we can add more info.
such as kyuubi.engine.url and kyuubi.engine.name.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala [hunk 3]
```diff
@@ -342,9 +351,12 @@ class ZookeeperDiscoveryClient(conf: KyuubiConf) extends DiscoveryClient {
...
+      s"serviceUri=$instance;version=${version.getOrElse(KYUUBI_VERSION)}" +
+        s"$extraInfo;${session}sequence=")
```
Comment (zwangsheng) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala:359]: we need `;` here to split version and extraInfo

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala (hunk 2 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala:96]: why extra is added? do we have an info method?

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/api/v1/dto/Engine.java [hunk 4]
```diff
@@ -40,14 +43,16 @@ public Engine(
       String sharelevel,
       String subdomain,
       String instance,
-      String namespace) {
+      String namespace,
+      Map<String, String> details) {
```
Comment (yaooqinn) [inline kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/api/v1/dto/Engine.java:47]: what does details stand for?

**Comment by yaooqinn** [comment]
2 general questions
 - why do need this?
 - why do we exposing in such a way?

**Comment by turboFei** [comment]
> why do need this?

We now support to list engines with kyuubi-admin command line.
And we need more insight for the engine information.

https://github.com/apache/incubator-kyuubi/pull/3680

> why do we expose it in this way?

**Comment by turboFei** [comment]
> why do we expose it in this way?

It is compatible with before.

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/api/v1/dto/Engine.java (hunk 4 above)
Comment (turboFei) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/api/v1/dto/Engine.java:47]: all the info saved in the engine discovery node.
version, sequence, engineId and so on.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala (hunk 2 above)
Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala:96]: > why extra is added? do we have an info method?

I did not get it.

**Comment by turboFei** [comment]
cc @yaooqinn

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala (hunk 2 above)
Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkTBinaryFrontendService.scala:96]: The extra is to distinguish w/ the service host/port information, if it's confusing, I suggest naming it attributes here.

**Review by pan3793 — APPROVED** [review]

**Comment by turboFei** [comment]
thanks for the review, merging to master

**Pull request closed without merging**